### DISCUSSION TITLE
How to support multiple certs in SSL_CTX

### DISCUSSION THREAD

**Discussion by YZLRYO** [openssl/openssl#22142] [Q&A] [answered]
Noob question. Is it possible to support multiple certs in one SSL_CTX?  Currently I'm using  use_certificate_chain_file to load cert. It's not supporting to load multiple certs. Is there an API to support?

**Comment by mattcaswell** [comment, marked as answer]
> Is it possible to support multiple certs in one SSL_CTX?

Yes you can have multiple leaf certs in one SSL_CTX. All such leaf certificates must be of a different type, e.g. RSA, ECDSA, ED25519 etc. You cannot have multiple leaf certs of the same type, e.g. you couldn't have 2 RSA certs.

> I'm using use_certificate_chain_file to load cert. It's not supporting to load multiple certs.

`SSL_CTX_use_certificate_chain_file` and `SSL_use_certificate_chain_file` enable you to supply a leaf cert and the intermediate CA certs in one go. The first cert in the file must be the leaf cert.

In order for the cert to work you must also supply the private key, e.g. via `SSL_CTX_use_private_key_file`.

You can repeat the calls to load a new cert with a different type.

For example to load an RSA cert and an ECDSA cert at the same time you might do something like this:

```` C
if (SSL_CTX_use_certificate_chain_file(ctx, "rsa-cert-and-chain.pem") <= 0)
    goto err;
if (SSL_CTX_use_PrivateKey_file(ctx, "rsa-key.pem", SSL_FILETYPE_PEM) <= 0)
    goto err;
if (SSL_CTX_use_certificate_chain_file(ctx, "ecdsa-cert-and-chain.pem") <= 0)
    goto err;
if (SSL_CTX_use_PrivateKey_file(ctx, "ecdsa-key.pem", SSL_FILETYPE_PEM) <= 0)
    goto err;
````

**Reply by YZLRYO** [reply]
Thanks for helping!
